scripts/irods/test/test_iquery.py

```diff
@@ -432,3 +432,48 @@ def test_rodsadmin_can_use_genquery2_to_find_collections_without_any_permissions
             self.admin.run_icommand(['ichmod', '-M', 'own', self.user.username, self.user.session_collection])
+
+    def test_genquery2_exposes_user_type_through_permission_related_columns__issue_8754(self):
+        # Create some data objects for the rodsadmin and rodsuser.
+        data_objects = ['data_object_0.8754', 'data_object_1.8754']
+        for name in data_objects:
+            self.user.assert_icommand(['itouch', name])
+            self.admin.assert_icommand(['itouch', name])
+
+        # Give the rodsuser permission to read the rodsadmin's session collection and one data object.
+        # This makes sure that the rodsadmin has some variation in the query results.
+        self.admin.assert_icommand(['ichmod', 'read_object', self.user.username, self.admin.session_collection])
+        self.admin.assert_icommand(['ichmod', 'read_object', self.user.username, data_objects[0]])
+
+        with self.subTest('rodsuser can use user type column for data objects'):
+            perms = [
+                [self.user.username, self.user.zone_name, 'own', 'rodsuser'],
+                [self.user.username, self.user.zone_name, 'own', 'rodsuser'],
+                [self.user.username, self.user.zone_name, 'read_object', 'rodsuser']
+            ]
+            json_string = json.dumps(perms, separators=(',', ':'))
+            query_string = "select DATA_ACCESS_USER_NAME, DATA_ACCESS_USER_ZONE, DATA_ACCESS_PERM_NAME, DATA_ACCESS_USER_TYPE order by DATA_ACCESS_PERM_NAME"
+            self.user.assert_icommand(['iquery', query_string], 'STDOUT', json_string)
+
+        with self.subTest('rodsuser can use user type column for collections'):
+            query_string = f"select COLL_ACCESS_USER_NAME, COLL_ACCESS_USER_ZONE, COLL_ACCESS_PERM_NAME, COLL_ACCESS_USER_TYPE where COLL_NAME = '{self.user.session_collection}'"
+            json_string = json.dumps([self.user.username, self.user.zone_name, 'own', 'rodsuser'], separators=(',', ':'))
+            self.user.assert_icommand(['iquery', query_string], 'STDOUT', [json_string])
+
+        with self.subTest('rodsadmin can use user type column for data objects'):
+            query_string = "select DATA_ACCESS_USER_NAME, DATA_ACCESS_USER_ZONE, DATA_ACCESS_PERM_NAME, DATA_ACCESS_USER_TYPE where DATA_NAME like '%.8754'"
+            _, out, _ = self.admin.assert_icommand(['iquery', query_string], 'STDOUT')
+            self.assertEqual(len(json.loads(out.strip())), 5)
+            self.assertIn(f'["{self.user.username}","{self.user.zone_name}","own","rodsuser"]', out)
+            self.assertIn(f'["{self.user.username}","{self.user.zone_name}","own","rodsuser"]', out)
+            self.assertIn(f'["{self.admin.username}","{self.admin.zone_name}","own","rodsadmin"]', out)
+            self.assertIn(f'["{self.admin.username}","{self.admin.zone_name}","own","rodsadmin"]', out)
+            self.assertIn(f'["{self.user.username}","{self.user.zone_name}","read_object","rodsuser"]', out)
```

Comment:
Let's put some space between the commas in these lines.

Also, just curious: Why are we not asserting a JSON string in this and the other rodsadmin sub-test (such as is being done in the rodsuser sub-tests) but rather a series of `assertIn`s? Is it because the ordering may differ?

Reply:
> Let's put some space between the commas in these lines.

That would cause the assertions to fail. GenQuery2 keeps the JSON compat to save space.

> Also, just curious: Why are we not asserting a JSON string in this and the other rodsadmin sub-test (such as is being done in the rodsuser sub-tests) but rather a series of assertIns? Is it because the ordering may differ?

Correct. It's because of ordering.

Reply:
I see. Resolving.